packages/docusaurus-plugin-google-gtag/src/utils/useGoogleAnalytics.ts

```diff
@@ -0,0 +1,25 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {useCallback} from 'react';
+
+export interface Event {
+  action: string;
+  event_category?: string;
+  event_label?: string;
+  value?: string;
+}
+
+export function useGoogleAnalytics(): {
+  sendEvent: (event: Event) => void;
+} {
+  const sendEvent = useCallback((event: Event) => {
+    window.gtag('event', event.action, event);
+  }, []);
+
+  return {sendEvent};
+}
```

Comment:
I'm not sure to understand the value of this abstraction 🤷‍♂️ 

What prevents someone to just do this directly in their code? (which would give them even more flexibility)

```js
function onClick() {
  const event = {
      action: 'some action',
      event_category: 'come category',
      event_label: 'some label',
      value:'some value'
  }

  window.gtag('event', event.action, event);
}
```

Reply:
They can, I'm just an old school OOO programmer and will always provide strong types. :)

Reply:
I can agree with that, but the types should rather be provided by ambient definitions from official gtag TS bindings preferably.

If they don't provide those, I'd rather create our own types only, and not add an extra runtime api/hook

For example:

```ts
interface Window {
  gtag: ...
}
```

Isn't this good enough to provide strong types for our users? Do we really need the extra client api/hook?

Reply:
> I can agree with that, but the types should rather be provided by ambient definitions from official gtag TS bindings preferably.
> 
> If they don't provide those, I'd rather create our own types only, and not add an extra runtime api/hook

There are gtag types published to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/gtag.js). Are you suggesting I get rid of the `Event` interface and replace it with something exported from there?

> For example:
> 
> ```ts
> interface Window {
>   gtag: ...
> }
> ```
> 
> Isn't this good enough to provide strong types for our users? Do we really need the extra client api/hook?

I couldn't get that approach to work with my code. Its possible I was doing something wrong, but I always got a syntax error that gtag does not exist on the Window object.

In fact, the reason that I don't reference the Event interface in `src/types.d.ts` is because if I try to use an import statement in that file I get the same syntax error. 

The `onClick` approach you suggested above does indeed work, but it requires using a `ts-ignore` statement because gtag is not global. Our internal code analyzers get very angry when we do stuff like that.

Reply:
PS - there is an unofficial [React gtag](https://github.com/idmadj/ga-gtag) package available. Its basically a port of the popular [ReactGA](https://github.com/react-ga/react-ga) plugin (that only works with old GA3 sites). 

I thought about going that route instead but it seemed like a much bigger lift that this approach.

Reply:
If we ever import a third-party package, it should be either the official one of the DT one.

This one looks appropriate to me:
- https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/gtag.js
- https://www.npmjs.com/package/@types/gtag.js

As far as I know adding a `@types/*` deps to the package should be enough for TS to find its ambient typedefs

If that's not the case you may be able to use `///` directive: 

https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html

```ts
/// <reference types="@types/gtag.js" />
```

---

> Our internal code analyzers get very angry when we do stuff like that.

On the Docusaurus repo we only care about providing a setup that TS agree with, not all the analyzers in the world 😅